specs/escrow.md

```diff
@@ -0,0 +1,60 @@
+# `SphinxEscrow` Contract Specification
+
+When deploying via the Sphinx DevOps Platform, the user has the option to pay for their deployments using ERC20 tokens on a single network. Before the deployment begins, the user must grant an allowance on an ERC20 token to the `SphinxEscrow` contract that is sufficient to cover the estimated cost of the deployment. The ERC20 token is then locked in the `SphinxEscrow` contract while the deployment is being executed. After the deployment is completed, the Sphinx backend calculates the exact cost of the deployment and transfers the remainder back to the user.
+
+The Sphinx team owns the `SphinxEscrow` contract.
+
+## Table of Contents
+
+- [Relevant files](#relevant-files)
+- [Use Case](#use-case)
+- [High-Level Invariants](#high-level-invariants)
+- [Function-Level Invariants](#function-level-invariants)
+- [Assumptions](#assumptions)
+
+## Relevant Files
+
+- The contract: [`SphinxEscrow.sol`](todo)
+- Unit tests: [`SphinxEscrow.t.sol`](todo)
+
+## Use Case
+
+### Allow Sphinx to accept crypto payments for the variable cost of deployments
+For Sphinx to accept crypto payments we must be able to guarantee that we will be paid for the cost of the deployment. Since the exact cost of a deployment cannot be exactly known ahead of time, we cannot charge for the deployment upfront. So instead, we estimate the cost of a deployment and require that the user transfer the estimated amount to an escrow contract while the deployment is being executed. After the deployment is completed, we calculate the exact cost and transfer the remaining funds back to the user.
+
+## High-Level Invariants
+- It must be possible to use the `SphinxEscrow` contract with any arbitrary ERC20 contract.
+- It must be possible to use the `SphinxEscrow` contract to transfer ERC20 tokens from an `owner` address to the `SphinxEscrow` contract if the `SphinxEscrow` contract has been granted an allowance.
+- It must *not* be possible to transfer ERC20 tokens from an `owner` address to any address other than the `SphinxEscrow` contract if the `SphinxEscrow` contract has been granted an allowance.
+- It must *only* be possible to transfer ERC20 tokens out of the `SphinxEscrow` contract to the original ERC20 `owner` address or an address with the `DEFAULT_ADMIN_ROLE` on the `SphinxEscrow` contract.
+
+## Function-Level Invariants
+
+#### `constructor(address _owner)`
+- Must grant the `DEFAULT_ADMIN_ROLE` to the specified owner address.
```

Comment:
It looks like we're using OpenZeppelin's `AccessControl` contract, but I only see the `DEFAULT_ADMIN_ROLE` used as a permission (in the `disburse` function). If we're planning on only having a single owner of this contract, we should use `Ownable` instead of `AccessControl`. If, instead, we're planning on allowing multiple addresses to withdraw funds, I think we should give them a specified role like we do in the `ManagedService` contract instead of giving them all the `DEFAULT_ADMIN_ROLE`. I think only one address should have the `DEFAULT_ADMIN_ROLE` because this address has permission to add or remove any other roles. This means that if we have multiple addresses with the `DEFAULT_ADMIN_ROLE` and we leak only one of their private keys, an attacker could remove all owners